lib/MySQL_Monitor.cpp

```diff
@@ -5816,5 +5891,1340 @@ bool DNS_Cache::empty() const {
+								read_only = 0;
+						}
+VALGRIND_ENABLE_ERROR_REPORTING;
+					}
+				}
+
+				rc = (*proxy_sqlite3_bind_int64)(statement, 5, read_only); ASSERT_SQLITE_OK(rc, mmsd->mondb);
+			} else {
+				proxy_error("mysql_fetch_fields returns NULL, or mysql_num_fields is incorrect. Server %s:%d . See bug #1994\n", mmsd->hostname, mmsd->port);
+				rc = (*proxy_sqlite3_bind_null)(statement, 5); ASSERT_SQLITE_OK(rc, mmsd->mondb);
+			}
+			mysql_free_result(mmsd->result);
+			mmsd->result = NULL;
+		} else {
+			rc = (*proxy_sqlite3_bind_null)(statement, 5); ASSERT_SQLITE_OK(rc, mmsd->mondb);
+		}
+		if (mmsd->result) {
+			// make sure it is clear
+			mysql_free_result(mmsd->result);
+			mmsd->result = NULL;
+		}
+		rc = (*proxy_sqlite3_bind_text)(statement, 6, mmsd->mysql_error_msg, -1, SQLITE_TRANSIENT); ASSERT_SQLITE_OK(rc, mmsd->mondb);
+		SAFE_SQLITE3_STEP2(statement);
+		rc = (*proxy_sqlite3_clear_bindings)(statement); ASSERT_SQLITE_OK(rc, mmsd->mondb);
+		rc = (*proxy_sqlite3_reset)(statement); ASSERT_SQLITE_OK(rc, mmsd->mondb);
+		(*proxy_sqlite3_finalize)(statement);
+
+		if (task_result == MySQL_Monitor_State_Data_Task_Result::TASK_RESULT_SUCCESS) {
+			MyHGM->read_only_action(mmsd->hostname, mmsd->port, read_only); // default behavior
+		} else {
+			char* error = NULL;
+			int cols = 0;
+			int affected_rows = 0;
+			SQLite3_result* resultset = NULL;
+			char* new_query = NULL;
+			SQLite3DB* mondb = mmsd->mondb;
+			new_query = (char*)"SELECT 1 FROM (SELECT hostname,port,read_only,error FROM mysql_server_read_only_log WHERE hostname='%s' AND port='%d' ORDER BY time_start_us DESC LIMIT %d) a WHERE read_only IS NULL AND SUBSTR(error,1,7) = 'timeout' GROUP BY hostname,port HAVING COUNT(*)=%d";
+			char* buff = (char*)malloc(strlen(new_query) + strlen(mmsd->hostname) + 32);
+			int max_failures = mysql_thread___monitor_read_only_max_timeout_count;
+			sprintf(buff, new_query, mmsd->hostname, mmsd->port, max_failures, max_failures);
+			mondb->execute_statement(buff, &error, &cols, &affected_rows, &resultset);
+			if (!error) {
+				if (resultset) {
+					if (resultset->rows_count) {
+						// disable host
+						proxy_error("Server %s:%d missed %d read_only checks. Assuming read_only=1\n", mmsd->hostname, mmsd->port, max_failures);
+						MyHGM->p_update_mysql_error_counter(p_mysql_error_type::proxysql, mmsd->hostgroup_id, mmsd->hostname, mmsd->port, ER_PROXYSQL_READ_ONLY_CHECKS_MISSED);
+						MyHGM->read_only_action(mmsd->hostname, mmsd->port, read_only); // N timeouts reached
+					}
+					delete resultset;
+					resultset = NULL;
+				}
+			} else {
+				proxy_error("Error on %s : %s\n", buff, error);
+			}
+			free(buff);
+		}
+	}
+
+	return true;
+}
+
+void MySQL_Monitor::monitor_read_only_async(SQLite3_result* resultset) {
+	assert(resultset);
+
+	std::vector<std::unique_ptr<MySQL_Monitor_State_Data>> mmsds;
+	mmsds.reserve(resultset->rows_count);
+	Monitor_Poll monitor_poll(resultset->rows_count);
```

Comment:
I have a suggestion for improving the the usage pattern of `Monitor_Poll` class, which I think it's a little odd right now regarding resource handling, as it could take much better advantage of the already used `std::unique_ptr`.

As I see, the usage pattern right now in the PR for `Monitor_Poll` is delegating the resource destruction of the created `MySQL_Monitor_State_Data` over a `std::vector<std::unique_ptr<MySQL_Monitor_State_Data>>`. This vector looks unused just for performing this resource handling, this creates a window for potential missuses of the class, since the manipulation of the data stored in the `std::unique_ptr<>` doesn't happen in the same scope in which the `std::unique_ptr<>` lives, instead, `get()` is used, invalidating the protection offered by the container. You know, I think this is normally fine when passing the pointer to a legacy function (system API, etc...) that is going to return in the same scope, but in this case, we are creating another container with the results of this `get()` which can have a completely different life-spam to the original `std::unique_ptr<>`.

What I suggest is to make this already constructed vector an argument for `Monitor_Poll` constructor, using `move semantics` for the initialization of an internal member. The `size()` member of the vector can be used for the same purpose as it's now used in the current constructor, initializing the real internal structure used for polling, and iterating it as it's done right now for calling: `monitor_poll.add((POLLIN|POLLOUT|POLLPRI), mmsd.get())`. There is no reason for even changing the internal logic for `add`, since its purpose will be the same, but since it won't have any external usage, we should probably declare it as private. With this change `Monitor_Poll` will be _well-formed_ resource wise since its construction, without worrying about scopes for the `MySQL_Monitor_State_Data`.

Reply:
Monitor Poll has aggregation type association with mmsd. Putting vector inside monitor poll class will change this association. Regarding mmsd life-span, this will never outlive vector, at-least in this current design.